File: 2021/python/63.unique-path-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if not obstacleGrid or len(obstacleGrid) == 0:
            return 0
        if obstacleGrid[-1][-1] == 1 or obstacleGrid[0][0] == 1:
            return 0
        rows = len(obstacleGrid)
        cols = len(obstacleGrid[0])
        
        dp = [[0] * cols for r in range(rows)]
        dp[0][0] = 1
        
        for r in range(rows):
            for c in range(cols):
                if obstacleGrid[r][c] == 0:
                    dp[r][c] += dp[r - 1][c] if r >= 1 else 0
                    dp[r][c] += dp[r][c - 1] if c >= 1 else 0
                    
        return dp[-1][-1]

File: 2021/java/63.unique-path-ii.java
class Solution {
    // dp: 
    // dp[i][j] -- represents max path can reach to cell (i,j)
    // 1. for any (i,j) == 1 (obstacle), reset dp[i][j] = 0
    // 2. for any(i,j) == 0, dp[i][j] = dp[i - 1][j] + dp[i][j - 1] (only right and down path)
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if (obstacleGrid == null || obstacleGrid.length == 0) return 0;
        int m = obstacleGrid.length;
        int n = obstacleGrid[0].length;
        int[][] dp = new int[m][n];
        // init first row
        for (int c = 0; c < n; c++) {
            if (obstacleGrid[0][c] == 1) break;
            dp[0][c] = 1;
        }
        // first col
        for (int r = 0; r < m; r++) {
            if (obstacleGrid[r][0] == 1) break;
            dp[r][0] = 1;
        }
        for (int r = 1; r < m; r++) {
            for (int c = 1; c < n; c++) {
                if (obstacleGrid[r][c] == 1) {
                    dp[r][c] = 0;
                } else {
                    dp[r][c] = dp[r][c - 1] + dp[r - 1][c];
                }
            }
        }
        return dp[m - 1][n - 1];
    }
}
